### PR TITLE
formula_auditor: refactor GCC dependency check into separate method

### PR DESCRIPTION
**Pull request by carlocab**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

The GCC dependency check is adding a couple of minutes to our
`tap_syntax` jobs. Let's fix that by moving the check into a separate
method so we can exclude it from `tap_syntax`.

### CODE REVIEW

File: Library/Homebrew/formula_auditor.rb [hunk 1]
```diff
@@ -865,14 +872,19 @@ def head_only?(formula)
...
-      return false unless Homebrew::SimulateSystem.simulating_or_running_on_macos?
+      odie "`#linux_only_gcc_dep?` works only on Linux!" if Homebrew::SimulateSystem.simulating_or_running_on_macos?
 
       formula_hash = formula.to_hash_with_variations
-      deps = formula_hash["dependencies"]
-      linux_deps = formula_hash.dig("variations", :x86_64_linux, "dependencies")
+      linux_deps = formula_hash["dependencies"]
+      variations_deps = []
+      formula_hash["variations"].map do |variation, data|
+        next if variation == :x86_64_linux
```
Comment (carlocab) [inline Library/Homebrew/formula_auditor.rb:881]: I don't think this is actually needed, but it doesn't hurt to play it safe, maybe?

File: Library/Homebrew/formula_auditor.rb [hunk 2]
```diff
@@ -865,14 +872,19 @@ def head_only?(formula)
     end
 
     def linux_only_gcc_dep?(formula)
```
Comment (carlocab) [inline Library/Homebrew/formula_auditor.rb:874]: Here's an unhandled edge-case (that currently isn't used by any formula), but is probably pre-existing:
```ruby
on_system :linux, macos: :big_sur_or_newer do
  depends_on "gcc"
end
```
This is falsely flagged as a Linux-only GCC dependency.

File: Library/Homebrew/formula_auditor.rb [hunk 3]
```diff
@@ -865,14 +872,19 @@ def head_only?(formula)
...
-      # TODO: Make this check work when running on Linux and not simulating macOS too.
-      return false unless Homebrew::SimulateSystem.simulating_or_running_on_macos?
+      odie "`#linux_only_gcc_dep?` works only on Linux!" if Homebrew::SimulateSystem.simulating_or_running_on_macos?
 
       formula_hash = formula.to_hash_with_variations
-      deps = formula_hash["dependencies"]
-      linux_deps = formula_hash.dig("variations", :x86_64_linux, "dependencies")
+      linux_deps = formula_hash["dependencies"]
+      variations_deps = []
+      formula_hash["variations"].map do |variation, data|
```
Comment (carlocab) [inline Library/Homebrew/formula_auditor.rb:880]: Maybe I need to be checking this `variations` hash for each of `MacOSVersions::SYMBOLS.keys`.

Comment (carlocab) [reply Library/Homebrew/formula_auditor.rb:880]: Oh, but doing this ignores the `arm64_*` keys.

File: Library/Homebrew/formula_auditor.rb [hunk 4]
```diff
@@ -865,14 +872,29 @@ def head_only?(formula)
...
+      return false if variations.blank?
+
+      # FIXME: This returns a false positive for formulae that do, for example:
+      # ```ruby
+      # on_system :linux, macos: :catalina_or_newer do
+      #   depends_on "gcc"
+      # end
+      # ```
+      variations_deps = []
+      formula_hash["variations"].each_value do |data|
```
Comment (carlocab) [inline Library/Homebrew/formula_auditor.rb:892]: I guess what I need to do is to check if all possible variations appear as keys in `formula_hash["variations"]`. If one is missing, then that means that variation has identical data to `:x86_64_linux`.

**Review by Rylan12 — APPROVED** [review]
This looks to me like it will work! Just to clarify, the edge case no longer exists in this most recent version, right?

**Comment by carlocab** [comment]
> This looks to me like it will work! Just to clarify, the edge case no longer exists in this most recent version, right?

Yes, it should no longer exist.

**Review by MikeMcQuaid — COMMENTED** [review]
Repeating from https://github.com/Homebrew/homebrew-test-bot/pull/821#issuecomment-1225801303:

> I'd rather we didn't do either of these `except` cases (`version,gcc_dependency`) and instead moved these to be either `--strict` or `--new-formula` or `--installed` audits instead.

**Comment by carlocab** [comment]
This isn't suitable for `--strict`, because we don't run `brew audit --strict` in CI, and I'm pretty sure this is intentional.

This isn't suitable for `--new-formula`, because most new GCC dependencies don't come from new formulae but from existing formulae that pick them up (and then infecting dependents as a consequence).

I'm not sure about `--installed` -- I don't think we run this in CI either. Should we do so? I don't know how that would work, though, since my understanding of `--installed` is that it just does `brew audit` but restricts it to installed formulae. We could probably run it, but that would result in: `brew audit`s of dependencies installed (maybe not bad, but can tend to increase the scope of PRs and result in unnecessary rebuilds), and multiple `brew audit`s of the same formula whenever a PR modifies more than one formula.

**Comment by MikeMcQuaid** [comment]
> I'm not sure about `--installed` -- I don't think we run this in CI either. Should we do so? I don't know how that would work, though, since my understanding of `--installed` is that it just does `brew audit` but restricts it to installed formulae. We could probably run it, but that would result in: `brew audit`s of dependencies installed (maybe not bad, but can tend to increase the scope of PRs and result in unnecessary rebuilds), and multiple `brew audit`s of the same formula whenever a PR modifies more than one formula.

Sorry, to clarify, I didn't mean `--installed` as much as "`brew audit` on a formula that has already been installed in test-bot" which is something we have for a few audits i.e. not part of a tap-wide `brew audit` by default.

**Comment by carlocab** [comment]
> i.e. not part of a tap-wide `brew audit` by default.

Ok, done.

File: Library/Homebrew/formula_auditor.rb [hunk 5]
```diff
@@ -415,6 +408,22 @@ def audit_conflicts
       end
     end
 
+    def audit_gcc_dependency
+      return unless @git
+      return unless @core_tap
```
Comment (MikeMcQuaid) [inline Library/Homebrew/formula_auditor.rb:413]: Would be good to have something similar for `audit_version` in `resource_auditor.rb` too but can wait for another PR.

**Review by MikeMcQuaid — APPROVED** [review]

File: Library/Homebrew/formula_auditor.rb (hunk 5 above)
Comment (carlocab) [reply Library/Homebrew/formula_auditor.rb:413]: Do we need it? I [benchmarked](https://github.com/carlocab/workflow-test/actions/runs/2928863489) `brew audit --tap=homebrew/core --only=version`, and it takes under 15 seconds (closer to 10 on Linux with `HOMEBREW_SIMULATE_MACOS_ON_LINUX`).

That's a long time, but not really by that much when doing `--except=version` already takes two minutes.

Comment (MikeMcQuaid) [reply Library/Homebrew/formula_auditor.rb:413]: @carlocab Fine with me! In that case, can you remove the `--except=version` in `brew test-bot`? Just want test bot and `brew audit` to be consistent by default here.

Comment (carlocab) [reply Library/Homebrew/formula_auditor.rb:413]: Actually, it even looks like most of that 10-15 seconds comes for free, as it's already done/cached elsewhere: https://github.com/carlocab/workflow-test/actions/runs/2928925233/attempts/1#summary-8023071550

Comment (carlocab) [reply Library/Homebrew/formula_auditor.rb:413]: https://github.com/Homebrew/homebrew-test-bot/pull/823
